src/Analyzers/Core/CodeFixes/NamespaceMatchFolder/AbstractChangeNamespaceToMatchFolderCodeFixProvider.CustomFixAllProvider.cs

```diff
@@ -0,0 +1,91 @@
+                    FixAllScope.Document when fixAllContext.Document is not null => await fixAllContext.GetDocumentDiagnosticsAsync(fixAllContext.Document).ConfigureAwait(false),
+                    FixAllScope.Project => await fixAllContext.GetAllDiagnosticsAsync(fixAllContext.Project).ConfigureAwait(false),
+                    FixAllScope.Solution => await GetSolutionDiagnosticsAsync(fixAllContext).ConfigureAwait(false),
+                    _ => default
+                };
+
+                if (diagnostics.IsDefaultOrEmpty)
+                    return null;
+
+                return new MyCodeAction(
+                    FixAllContextHelper.GetDefaultFixAllTitle(fixAllContext),
+                    cancellationToken => FixAllByDocumentAsync(
+                        fixAllContext.Project.Solution,
+                        diagnostics,
+                        cancellationToken));
+
+                static async Task<ImmutableArray<Diagnostic>> GetSolutionDiagnosticsAsync(FixAllContext fixAllContext)
+                {
+                    var diagnostics = ImmutableArray.CreateBuilder<Diagnostic>();
+
+                    foreach (var project in fixAllContext.Solution.Projects)
+                    {
+                        var projectDiagnostics = await fixAllContext.GetAllDiagnosticsAsync(fixAllContext.Project).ConfigureAwait(false);
+                        diagnostics.AddRange(projectDiagnostics);
+                    }
+
+                    return diagnostics.ToImmutable();
+                }
+            }
+
+            private static async Task<Solution> FixAllByDocumentAsync(
+                Solution solution,
+                ImmutableArray<Diagnostic> diagnostics,
+                CancellationToken cancellationToken)
+            {
+                // Use documentId instead of tree here because the
+                // FixAsync call can modify more than one document per call. The
+                // important thing is that the fix works on fixing the namespaces in a single document,
+                // but references in other documents will be updated to be correct. Id will remain
+                // across this mutation, but lookup via SyntaxTree directly will not work because
+                // the tree won't be the same.
+                var documentIdToDiagnosticsMap = diagnostics
+                    .GroupBy(diagnostic => diagnostic.Location.SourceTree)
+                    .Where(group => group.Key is not null)
+                    .ToImmutableDictionary(group => solution.GetRequiredDocument(group.Key!).Id, group => group.ToImmutableArray());
+
+                var newSolution = solution;
+
+                foreach (var (documentId, diagnosticsInTree) in documentIdToDiagnosticsMap)
+                {
+                    var document = newSolution.GetRequiredDocument(documentId);
+
+                    newSolution = await FixAllInDocumentAsync(document, diagnosticsInTree, cancellationToken).ConfigureAwait(false);
```

Comment:
looking at this, is there a reason you don't just use DocumentBasedFixAllProvider?  it already exposes an api where you just are given the individual docs and diagnostics, and you just have to worry about fixing one doc at a time.

Reply:
oh.  nevermind.  your fixes are updating lots of code in teh solution.

however, i think you need to add support for showing progress.  This could have to fixup *lots* of files, and each of those renames will be *Expensive*.  we need to show progress so users can know how long they have to wait in a fix all.  i can help you with this.

Reply:
Adding an IProgressTracker for this